### PR TITLE
lightning: allocate autoid globally when increment-import is enabled

### PR DESCRIPTION
**Pull request by gozssky**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/35904

Problem Summary:

See https://github.com/pingcap/tidb/issues/35904.

### What is changed and how it works?

When `increment-import` is enabled, this PR will allocate autoid globally before restoring starts.  Thus we can import data to a partition that won't conflict with normal read/write operations to other partitions.

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/tests/lightning_partition_incremental/run.sh [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+# Must wait for pid explicitly. Otherwise, we can't detect the process is exited with error.
+wait "$pid1" "$pid2"
+
+grep -F 'checksum pass' "$LOG_FILE1" | grep -Fq 'auto_rowid'
+grep -F 'checksum pass' "$LOG_FILE1" | grep -Fq 'non_pk_auto_inc'
+
+run_sql 'admin check table incr.auto_rowid'
+run_sql 'admin check table incr.non_pk_auto_inc'
+
+run_sql 'select count(*) from incr.auto_rowid'
```
Comment (glorv) [inline br/tests/lightning_partition_incremental/run.sh:38]: Please also check that the are no tidb_rowid duplicate across all partitions.

**Review by glorv — COMMENTED** [review]
rest LGTM

File: br/pkg/lightning/restore/meta_manager.go [hunk 2]
```diff
@@ -244,44 +244,21 @@ func (m *dbTableMetaMgr) AllocTableRowIDs(ctx context.Context, rawRowIDMax int64
...
-					}
-				}
-				if len(autoIDField) == 0 && common.TableHasAutoRowID(m.tr.tableInfo.Core) {
-					autoIDField = model.ExtraHandleName.L
-				}
-				if len(autoIDField) == 0 {
-					return common.ErrAllocTableRowIDs.GenWithStack("table %s contains auto increment id or _tidb_rowid, but target field not found", m.tr.tableName)
+			if needAutoID {
+				// maxRowIDMax is the max row_id that other tasks has allocated, we need to rebase the global autoid base first.
+				if err := rebaseGlobalAutoID(ctx, maxRowIDMax, m.tr.kvStore, m.tr.dbInfo.ID, m.tr.tableInfo.Core); err != nil {
```
[the remark this answers is not in the document]
Comment (gozssky) [reply br/pkg/lightning/restore/meta_manager.go:249]: I'm not sure if there are any users who rely on the `auto_increment` in table schema.🤔

When we really need `increment-import`, it mostly means the table already contains data and the schema has been created before. (except for parallel-import)

File: br/tests/lightning_partition_incremental/run.sh (hunk 1 above)
Comment (gozssky) [reply br/tests/lightning_partition_incremental/run.sh:38]: Fixed in https://github.com/pingcap/tidb/pull/35905/commits/729a523818e1c6db14923811196dde384ca58163.

**Review by buchuitoudegou — APPROVED** [review]
lgtm

**Comment by glorv** [comment]
I think the logic is correct. But there is behavior change as after this PR the auto_id is allocated from the table allocate base instead of 0. As the table schema export from mydumper/dumpling contains the auto_increment attribute of the current value, we may need to reset it in lightning at creating table schema.
/cc @lichunzhu what do you think?

**Comment by lichunzhu** [comment]
> I think the logic is correct. But there is behavior change as after this PR the auto_id is allocated from the table allocate base instead of 0. As the table schema export from mydumper/dumpling contains the auto_increment attribute of the current value, we may need to reset it in lightning at creating table schema.

As long as the data loaded into downstream have correct auto increment IDs, I think these gaps don't matter. Dumpling will also generate gap when we dump sequences.

**Comment by gozssky** [comment]
> But there is behavior change as after this PR the auto_id is allocated from the table allocate base instead of 0.

@glorv There may be no big behavior change after this PR. When `increment-import` is enabled, we get the `next_row_id` as `rowIDbase` which should consider the `auto_increment` in the table schema.

**Review by glorv — APPROVED** [review]

**Pull request merged**